core/src/Streamly/Internal/FileSystem/Posix/ReadDir.hsc

```diff
@@ -138,41 +257,64 @@ isMetaDir dname = do
                 then return True
                 else return False
 
-{-# NOINLINE lstatDname #-}
-lstatDname :: PosixPath -> Ptr CChar -> IO (Bool, Bool)
-lstatDname parent dname = do
+data GStatRes
+    = GSIsMetaDir
+    | GSIsRegDir
+    | GSIsNotDir
+    | GSIgnoreError Errno
+
+{-# NOINLINE gstatDname #-}
+gstatDname
+    :: ReadOptions -> PosixPath -> Ptr CChar -> IO GStatRes
+gstatDname conf parent dname = do
     isMeta <- liftIO $ isMetaDir dname
     if isMeta
-    then pure (True, True)
+    then pure GSIsMetaDir
     else do
         -- XXX We can create a pinned array right here since the next call pins
         -- it anyway.
         path <- appendCString parent dname
         Array.asCStringUnsafe (Path.toChunk path) $ \cStr -> do
-            res <- c_lstat_is_directory cStr
+            res <-
+                if _followSymlinks conf
+                then statWith c_stat cStr
+                else statWith c_lstat cStr
             case res of
-                x | x == 1 -> pure (True, False)
-                x | x == 0 -> pure (False, False)
-                -- XXX Need to check if and how we should handle some errors
-                -- like EACCES.
-                _ -> throwErrno "checkIfDirectory"
+                Left errno ->
+                    if    (errno == eNOENT && _ignoreNonExistingFiles conf)
+                       || (errno == eACCES && _ignoreInAccessibleFiles conf)
+                       || (errno == eLOOP && _ignoreSymlinkLoopErrors conf)
```

Comment:
this is technically not efficient - even if errno was eNOENT  it cannot be eACCES but you might still check whether it is eACCES or not.

Reply:
`||` is lazy in nature right?
`True || x` = True

Reply:
That is not the point, the point is - If errno == eNOENT is True and _ignoreNonExistingFiles conf is False we still check the other cases but we should not.